water/lib/Dialect/Wave/Transforms/LowerReadWriteOps.cpp

```diff
@@ -371,7 +371,14 @@ static FailureOr<MemAccessInfo> createMemoryIndicesAndMask(
   // This logic operates after the "expansion" pass that usually unrolls all
   // tensor dimensions but one so the extent along them is 1. The dimension
   // with non-unit extent is considered to be vectorized.
-  DictionaryAttr indexDict = op.getIndexAttr();
+
+  // Read/Write ops only carry a single index expression: the first (and only)
+  // dictionary inside the array attribute.
+  ArrayAttr indexArr = op.getIndexAttr();
+  if (!indexArr || indexArr.size() != 1)
```

Comment:
Normally, I'd expect a verifier to check that, when the index array is present, there's exactly one element in it. And the second check would turn into an assert here and a diagnostic on the op, not a silent failure-to-match.